drivers/adc/ad7616/iio_ad7616.c

```diff
@@ -0,0 +1,310 @@
+		   intptr_t id)
+{
+	int ret;
+	uint8_t index = channel->ch_num < 8 ? channel->ch_num : channel->ch_num - 8;
+	struct ad7616_conversion_result res[8];
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	// TODO: we call this on every call, fetching all 16 inputs. Doesn't look smart
+	ret = ad7616_read_data_serial(dev, res, 1);
+	if (ret)
+		return ret;
+
+	if (channel->ch_num < 8)
+		return sprintf(buf, "%i", res[index].channel_a);
+
+	return sprintf(buf, "%i", res[index].channel_b);
+}
+
+static int get_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	enum ad7616_range range = (ch < AD7616_VB0) ? dev->va[ch] : dev->vb[ch -
+				  AD7616_VB0];
+
+	switch (range) {
+	case AD7616_10V:
+		return sprintf(buf, "10");
+	case AD7616_5V:
+		return sprintf(buf, "5");
+	case AD7616_2V5:
+		return sprintf(buf, "2.5");
+	default:
+		return -EINVAL;
+	}
+}
+
+static int set_scale(void *device,
+		     char *buf,
+		     uint32_t len,
+		     const struct iio_ch_info *channel,
+		     intptr_t id)
+{
+	int ret;
+	int val;
+	enum ad7616_range range;
+	enum ad7616_ch ch = (channel->ch_num < 8) ? channel->ch_num : AD7616_VB0 +
+			    channel->ch_num - 8;
+	struct ad7616_iio_dev *iio_dev = device;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+
+	ret = sscanf(buf, "%u", &val);
+	if (ret != 1)
+		return -EINVAL;
+
+	switch (val) {
+	case 10:
+		range = AD7616_10V;
+		break;
+	case 5:
+		range = AD7616_5V;
+		break;
+	case 2:
+		range = AD7616_2V5;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	return ad7616_set_range(dev, ch, range);
+}
+
+/**
+ * @brief	Read buffer data corresponding to AD7616 IIO device
+ * @param	iio_dev_data[in] - Pointer to IIO device data structure
+ * @return	0 in case of success, negative error code otherwise
+ */
+static int32_t iio_ad7616_submit_buffer(struct iio_device_data *iio_dev_data)
+{
+	struct ad7616_conversion_result results[8];
+	struct ad7616_iio_dev *iio_dev = iio_dev_data->dev;
+	struct ad7616_dev *dev = iio_dev->ad7616_dev;
+	uint16_t data[iio_dev_data->buffer->bytes_per_scan / 2];
+	int32_t ret;
+	uint32_t i;
+	uint32_t j;
+	uint32_t k;
+
+	for (i = 0; i < iio_dev_data->buffer->samples; i++) {
+		ret = ad7616_read_data_serial(dev, results, 1);
+		if (ret)
+			return ret;
+
+		for (j = 0, k = 0; j < 16; j++) {
+			if (!(iio_dev_data->buffer->active_mask & NO_OS_BIT(j)))
+				continue;
+
+			if (j < 8)
+				data[k++] = results[j].channel_a;
+			else
+				data[k++] = results[j - 8].channel_b;
+		}
+
+		ret = iio_buffer_push_scan(iio_dev_data->buffer, &data);
+		if (ret)
+			return ret;
+	}
+
+	return 0;
+}
+
+static struct scan_type ad7616_iio_scan_type = {
+	.sign = 's',
+	.realbits = CHN_STORAGE_BITS,
```

Comment:
Reading the documentation of `realbits` in `iio/iio_types.h` ("Number of valid bits of data") it looks to me that it should be `16` as there is `16` bits (2 bytes). Same for `storagebits` as there is no padding.

![image](https://github.com/analogdevicesinc/no-OS/assets/15075443/27af7889-5d04-4f69-a458-4de9827d85e9)

Reply:
so first let's discuss what data is added in the buffer.
please correct me if I am wrong. there are 16 channels available in the iio driver. the output value of a channel is either 0 or 1. these values should be added in the buffer based on the scan_mask.

the scan type is actually only used for buffering.. it shows how the channel data is added in the buffer. it does not have anything to do with the way data is retrieved from AD7616.

seeing how the data added in the buffer is only one bit per channel, the realbits should be 1 (because we are referring to the scan_type of the channel). now, the storage bits can really be any multiple of a byte, but it does not make sense to use something higher than a byte, because that data is not useful.

so to sum up, the data in the buffer should contain the data based on the scan_mask in the scan_type format.

now let's take a look at ad7616_read_data_serial and how the data is added in the result.
values for channel 0 -> 7 are found in bits 0 - bits 7 from res.channel_a
values for channel 8 -> 15 are found in bits 0 - bits 7 from res.channel_b

with this in mind, it looks like the get raw and the submit APIs should be rewritten
here is my suggestion for get raw:
```
int ret;
	uint8_t index = channel->ch_num < 8 ? channel->ch_num : channel->ch_num - 8;
	struct ad7616_conversion_result res;
	struct ad7616_iio_dev *iio_dev = device;
	struct ad7616_dev *dev = iio_dev->ad7616_dev;

	// TODO: we call this on every call, fetching all 16 inputs. Doesn't look smart
	ret = ad7616_read_data_serial(dev, res, 1);
	if (ret)
		return ret;

	if (channel->ch_num < 8)
                // will return 0 is the bit according to the channel is 0, and 1 if the bit is 1
		return sprintf(buf, "%i", !!(res.channel_a & NO_OS_BIT(index)));

	return sprintf(buf, "%i", !!(res.channel_b & NO_OS_BIT(index)));
```

Reply:
> there are 16 channels available in the iio driver. the output value of a channel is either 0 or 1.

Yes there is 16 channels. However each channel is 16 bits wide. As stated by the [datasheet](https://www.analog.com/media/en/technical-documentation/data-sheets/AD7616.pdf):
```
The AD7616 is a 16-bit, DAS that supports dual simultaneous sampling of 16 channels
```
In find the quote a bit confusing but in the block diagram you can see that there is a MUX for A0 to A7 (same for B0..B7) and the output goes to a 16bit SAR (one for A and one for B). There is 2 data path returning 16 bit each, A in red, B in green:
![image](https://github.com/analogdevicesinc/no-OS/assets/15075443/96b43a01-2a90-4222-9f3a-244897b30f30)

The Linux driver also seem to agree that a channel returns 16 bits: https://elixir.bootlin.com/linux/latest/source/drivers/iio/adc/ad7606.h#L22

The input channel for both channel A and B is selected via the "Channel Register" (see page 43). In burst mode as configured here, each time a conversion is requested the device is sending A0[0..16], B0[0..16], A1[0..16], B1[0..16], ..., A7[0..16], B7[0..16] (all 16 input channels)

Reply:
oh, alright. it all makes sense now. I apologize for the misunderstanding.

Reply:
No problem I got confuse too the first time I read the datasheet :D